WalletWasabi/EventSourcing/InMemoryEventRepository.cs

```diff
@@ -0,0 +1,216 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading.Tasks;
+using WalletWasabi.Exceptions;
+using WalletWasabi.Helpers;
+using WalletWasabi.Interfaces.EventSourcing;
+
+namespace WalletWasabi.EventSourcing
+{
+	/// <summary>
+	/// Thread safe without locks in memory event repository implementation
+	/// </summary>
+	public class InMemoryEventRepository : IEventRepository
+	{
+		protected static readonly IReadOnlyList<WrappedEvent> EmptyResult = ImmutableList<WrappedEvent>.Empty;
+		protected static readonly IReadOnlyList<string> EmptyIds = ImmutableList<string>.Empty;
+
+		private readonly ConcurrentDictionary
+			// aggregateType
+			<string,
+			ConcurrentDictionary
+				// aggregateId
+				<string,
+				(
+					// SequenceId of the last event of this aggregate
+					long TailSequenceId,
+
+					// Ordered list of events
+					ImmutableList<WrappedEvent> Events
+				)>> _aggregatesEventsBatches = new();
+
+		private readonly ConcurrentDictionary
+			// aggregateType
+			<string,
+			(
+				// Index of the last aggregateId in this aggregateType
+				long TailIndex,
+
+				// List of aggregate Ids in this aggregateType
+				ImmutableSortedSet<string> Ids
+			)> _aggregatesIds = new();
+
+		public Task AppendEventsAsync(
+			string aggregateType,
+			string aggregateId,
+			IEnumerable<WrappedEvent> wrappedEvents)
+		{
+			Guard.NotNullOrEmpty(nameof(aggregateType), aggregateType);
+			Guard.NotNullOrEmpty(nameof(aggregateId), aggregateId);
+			Guard.NotNull(nameof(wrappedEvents), wrappedEvents);
+			var wrappedEventsList = wrappedEvents.ToList().AsReadOnly();
+			if (wrappedEventsList.Count <= 0)
+			{
+				return Task.CompletedTask;
+			}
+			var firstSequenceId = wrappedEventsList[0].SequenceId;
+			var lastSequenceId = wrappedEventsList[^1].SequenceId;
+			if (firstSequenceId <= 0)
+			{
+				throw new ArgumentException("First event sequenceId is not natural number.", nameof(wrappedEvents));
+			}
+			if (lastSequenceId <= 0)
+			{
+				throw new ArgumentException("Last event sequenceId is not natural number.", nameof(wrappedEvents));
+			}
+			if (lastSequenceId - firstSequenceId + 1 != wrappedEventsList.Count)
+			{
+				throw new ArgumentException("Event sequence ids are out of whack.", nameof(wrappedEvents));
```

Comment:
Whack is a rarely used word. I reviewed this with Adam and he was blocked by it.

Reply:
```suggestion
				throw new ArgumentException("Event sequence ids are inconsistent.", nameof(wrappedEvents));
```